pkg/ruler/ruler.go

```diff
@@ -188,66 +162,50 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d
 // Stop stops the Ruler.
 // Each function of the ruler is terminated before leaving the ring
 func (r *Ruler) Stop() {
+	close(r.done)
+	<-r.terminated
```

Comment:
this line waits on the terminated channel to stop before proceeding.  this channel is closed by the `run()` method which calls `loadRules()` in a loop.  `loadRules()` and `Stop()` both use the `userManagerMtx` to lock each other out of the `userManagers` map.  Is this necessary?  If `Stop()` is guaranteed to run independent of `loadRules()` I think dropping the mutex for clarity would be good.

Reply:
I think in this case it would be best to maintain the mutex. Especially because it will likely be needed to implement the `/api/v1/rules` prometheus endpoint which returns the following json:

```json
{
    "status": "success",
    "data": {
        "groups": [
            {
                "name": "alertmanager.rules",
                "file": "/etc/prometheus/alerts.rules",
                "rules": [
                    {
                        "name": "AlertmanagerFailedReload",
                        "query": "alertmanager_config_last_reload_successful{job=\"default/alertmanager\"} == 0",
                        "duration": 600,
                        "labels": {
                            "severity": "warning"
                        },
                        "annotations": {
                            "message": "Reloading Alertmanager's configuration has failed for {{ $labels.namespace }}/{{ $labels.pod}}."
                        },
                        "alerts": [],
                        "health": "ok",
                        "type": "alerting"
                    },
...
```

In order to get the health of indvidual rules we are going to have to allow for concurrent access to the userManager map.